Lib/exception_group.py

```diff
@@ -0,0 +1,215 @@
+        of self for which condition(e) returns True and False, respectively.
+        match and rest have the same nested structure as self, but empty
+        sub-exceptions are not included.
+
+        condition: BaseException --> Boolean
+        """
+        match, rest = [], []
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup): # recurse
+                e_match, e_rest = e.project(condition)
+                if e_match:
+                    match.append(e_match)
+                if e_rest:
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                    e_match, e_rest = e, None
+                else:
+                    rest.append(e)
+        return (ExceptionGroup(match, tb=self.__traceback__),
+                ExceptionGroup(rest, tb=self.__traceback__))
+
+    def split(self, type):
+        """ Split an ExceptionGroup to extract exceptions of type E
+
+        type: An exception type
+        """
+        return self.project(lambda e: isinstance(e, type))
+
+    def subgroup(self, keep):
+        """ Split an ExceptionGroup to extract only exceptions in keep
+
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
+        """ returns the traceback of a single exception
+
+        If exc is in this exception group, return its
+        traceback as a list of frames. Otherwise, return None.
+        """
+        if exc not in self:
+            return None
+        result = []
+        e = self.subgroup([exc])
+        while e:
+            tb = e.__traceback__
+            while tb:
+                result.append(tb.tb_frame)
+                tb = tb.tb_next
+            if isinstance(e, ExceptionGroup):
+                assert len(e.excs) == 1 and exc in e
+                e = e.excs[0]
+            else:
+                assert e is exc
+                e = None
+        return result
+
+    def push_frame(self, frame):
+        import types
+        self.__traceback__ = types.TracebackType(
+            self.__traceback__, frame, 0, 0)
+
+    @staticmethod
+    def render(exc, tb=None, indent=0):
+        output = []
+        output.append(f"{exc}")
+        tb = tb or exc.__traceback__
+        while tb and not isinstance(tb, TracebackGroup):
+            output.append(f"{' '*indent} {tb.tb_frame}")
+            tb = tb.tb_next
+        if isinstance(exc, ExceptionGroup):
+            tbg = exc.__traceback_group__
+            assert isinstance(tbg, TracebackGroup)
+            indent += 4
+            for e in exc.excs:
+                t = tbg.tb_next_map[id(e)]
+                output.append('---------------------------------------')
+                output.extend(ExceptionGroup.render(e, t, indent))
+        for l in output:
+            print(l)
```

Comment:
Debug print?

Reply:
I guess that confused me -- it painstakingly constructs a list of lines and returns it, *and* also prints it. :-)